File: stvs/src/test/java/edu/kit/iti/formal/stvs/model/config/HistoryTest.java
package edu.kit.iti.formal.stvs.model.config;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

/**
 * @author Benjamin Alt
 */
public class HistoryTest {
  private History history;

  @Before
  public void setUp() {
    history = new History();
  }

  @Test
  public void testBufferSize() {
    assertEquals(0, history.getFilenames().size());
    for (int i = 0; i < History.HISTORY_DEPTH * 2; i++) {
      history.addFilename("filePath" + i);
    }
    assertEquals(History.HISTORY_DEPTH, history.getFilenames().size());
  }

  @Test
  public void testConstructor() {
    ArrayList<String> filePaths = new ArrayList<String>();
    filePaths.add("CodeOne");
    filePaths.add("CodeTwo");
    filePaths.add("SpecOne");
    history = new History(filePaths);
    List<String> filePathsAfter = history.getFilenames();
    assertEquals(filePathsAfter.get(0), "CodeOne");
    assertEquals(filePathsAfter.get(1), "CodeTwo");
    assertEquals(filePathsAfter.get(2), "SpecOne");
  }

  @Test(expected = IllegalArgumentException.class)
  public void testConstructorException() {
    ArrayList<String> codePaths = new ArrayList<String>();
    for (int i = 0; i < History.HISTORY_DEPTH * 2; i++) {
      codePaths.add("SomeCode" + i);
    }
    history = new History(codePaths);
  }

  @Test
  public void testAddSpecFilename() {
    history.addFilename("someSpec");
    assertEquals(history.getFilenames().get(0), "someSpec");
  }

  @Test
  public void testSetAll() {
    testConstructor();
    History clone = new History();
    clone.setAll(history);
    assertEquals(history, clone);
  }

  @Test
  public void testEquals() {
    testConstructor();
    ArrayList<String> filePaths = new ArrayList<String>();
    filePaths.add("CodeOne");
    filePaths.add("CodeTwo");
    filePaths.add("SpecOne");
    History identical = new History(filePaths);
    assertEquals(history, identical);
    assertNotEquals(null, history);
    assertEquals(history, history);
    identical.getFilenames().add("Another filename!");
    assertNotEquals(history, identical);
  }

  @Test
  public void testHashCode() {
    testConstructor();
    ArrayList<String> filePaths = new ArrayList<String>();
    filePaths.add("CodeOne");
    filePaths.add("CodeTwo");
    filePaths.add("SpecOne");
    History identical = new History(filePaths);
    assertEquals(history.hashCode(), identical.hashCode());
    assertEquals(history.hashCode(), history.hashCode());
    identical.getFilenames().add("Another filename!");
    assertNotEquals(history.hashCode(), identical.hashCode());
  }
}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/model/config/HistoryTest.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.config

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertFailsWith

/**
 * @author Benjamin Alt
 */
class HistoryTest {
    private var history: History? = null

    @BeforeEach
    fun setUp() {
        history = History()
    }

    @Test
    fun testBufferSize() {
        Assertions.assertEquals(0, history!!.filenames.size.toLong())
        for (i in 0 until History.HISTORY_DEPTH * 2) {
            history!!.addFilename("filePath$i")
        }
        Assertions.assertEquals(History.HISTORY_DEPTH.toLong(), history!!.filenames.size.toLong())
    }

    @Test
    fun testConstructor() {
        val filePaths = ArrayList<String>()
        filePaths.add("CodeOne")
        filePaths.add("CodeTwo")
        filePaths.add("SpecOne")
        history = History(filePaths)
        val filePathsAfter: List<String> = history!!.filenames
        Assertions.assertEquals(filePathsAfter[0], "CodeOne")
        Assertions.assertEquals(filePathsAfter[1], "CodeTwo")
        Assertions.assertEquals(filePathsAfter[2], "SpecOne")
    }

    @Test // (expected = IllegalArgumentException::class)
    fun testConstructorException() {
        val codePaths = ArrayList<String>()
        for (i in 0 until History.HISTORY_DEPTH * 2) {
            codePaths.add("SomeCode$i")
        }
        assertFailsWith<IllegalArgumentException> {
            history = History(codePaths)
        }
    }

    @Test
    fun testAddSpecFilename() {
        history!!.addFilename("someSpec")
        Assertions.assertEquals(history!!.filenames[0], "someSpec")
    }

    @Test
    fun testSetAll() {
        testConstructor()
        val clone = History()
        clone.setAll(history!!)
        Assertions.assertEquals(history, clone)
    }

    @Test
    fun testEquals() {
        testConstructor()
        val filePaths = ArrayList<String>()
        filePaths.add("CodeOne")
        filePaths.add("CodeTwo")
        filePaths.add("SpecOne")
        val identical = History(filePaths)
        Assertions.assertEquals(history, identical)
        Assertions.assertNotEquals(null, history)
        Assertions.assertEquals(history, history)
        identical.filenames.add("Another filename!")
        Assertions.assertNotEquals(history, identical)
    }

    @Test
    fun testHashCode() {
        testConstructor()
        val filePaths = ArrayList<String>()
        filePaths.add("CodeOne")
        filePaths.add("CodeTwo")
        filePaths.add("SpecOne")
        val identical = History(filePaths)
        Assertions.assertEquals(history.hashCode().toLong(), identical.hashCode().toLong())
        Assertions.assertEquals(history.hashCode().toLong(), history.hashCode().toLong())
        identical.filenames.add("Another filename!")
        Assertions.assertNotEquals(history.hashCode().toLong(), identical.hashCode().toLong())
    }
}